Add part 2 of the tutorial: CLI Contact Manager (#86)
davenverse/shellfish: docs/tutorial/creating_cli.md

# Creating a Contact Manager CLI

> At this point, you may be very advance in the use of Shellfish and Cats Effect. That is why in this tutorial, we will create a simple CLI application that manages contacts. The application will allow users to add, list, update, and delete contacts.

Reviewer: ```suggestion
Moving on from the basics, in this tutorial, we will create a simple CLI application that manages contacts. The application will allow users to add, list, update, and delete contacts.
```

```bash
$ cm add 
> Enter username: contact_username
> First Name: contact_name
> Last Name: contact_name
> Phone: contact_phone
> Email: contact_email
```

```bash
$ cm get 
> ----- contact_username -----
> First Name:    contact_name
> Last Name:     contact_name
> Phone:         contact_phone
> Email:         contact_email
```

```bash
$ cm update contact_username --last-name new_contact_name
> Contact contact_username updated successfully
```

Now that we know what we want to build, let us start by creating the project structure.

## Project Structure
For this project, we will use the following project structure:

```
examples/
└── src
    └── main
        └── scala
            └── contacts
                ├── app
                ├── cli
                ├── core
                └── domain
```

- `app`: This package will contain the main entry point of the application.
- `cli`: This package will contain the CLI logic such as prompt parsing and command execution.
- `core`: This package will contain the business logic of the application, here is where we will use Shellfish!

> - `domain`: This package will contain the domain model of the application, such as all the types.

Reviewer: ```suggestion
- `domain`: This package will contain important business logic that `app`, `cli` and `core` will use. 
```

Let us first start by defining our domain.

## Domain
Our domain will be straightforward, we will have a `Contact` case class that will represent a contact. 

> ```scala

Reviewer: Could you add the path of the file where the code belongs to for each codeblock? Like in a comment
```
./src/main/contacts/domain/Types.scala
```

type Username    = String
type Name        = String
type PhoneNumber = String
type Email       = String

case class Contact(
  username: Username,
  firstName: Name,
  lastName: Name,
  phoneNumber: PhoneNumber,
  email: Email
)
``` 

> The type aliases are just for better readability, but you can use refined types provided by nice libraries like [Iron Types](https://github.com/Iltotore/iron), but for the sake of simplicity, we will use plain strings.

Reviewer: Besides readability, type aliases provide a way to not accidentally pass in the wrong string in a function that expects several strings. I think mentioning that instead of readability is better(imo)

We will also need a function to display the contact in a nice format. If you came from Java, this function is known to `toString()`. But in Scala, we use `show` for this purpose.

```scala 3
case class Contact( ... ):
  def show: String = s"""
    |----- $username -----
    |First Name:    $firstName
    |Last Name:     $lastName
    |Phone:         $phoneNumber
    |Email:         $email
  """.stripMargin
```

> We also will need a custom error type that we will use in case, for example, if we want to add a contact that already exists.

Reviewer: ```suggestion
We will also define a custom error type that gets thrown when a contact already exists.
```

      case None =>
        lines.isEmpty
          .pure[IO]
          .ifM( // (3)
            bookPath.write(encodeContact(contact)).as(contact.username),  // (4)
            bookPath.appendLine(encodeContact(contact)).as(contact.username)
          )
          
end addContact
```

First, we read all of our lines, line by line using the `FilesOs#readLines` method (1). Then, we check if the contact already exists in the file checking if the username is contained (2). If it does, we raise an error using the `raiseError` method. If it doesn't, we check if the file is empty (3) using the `ifM`; this method basically checks if the boolean inside the `IO` is true or false and it executes an effect depending on the condition. If it is, we write the contact to the file using the `write` method. If it's not, we append the contact to the file using the `appendLine` method (4).

Now that we have the `addContact` function implemented, we can move to the `removeContact` function:

```scala 3
override def removeContact(username: Username): IO[Unit] =
  for
    contacts <- bookPath.readLines
    _ <- bookPath.writeLines(
      contacts.filterNot(_.startsWith(username))
    )
  yield ()
```

> The explanation is straightforward. We load all the contacts from the file, and then we filter the ones that don't match the username we want to remove. Finally, we write the filtered contacts back to the file.

Reviewer: ```suggestion
Here we load all the contacts from the file, and then we filter the ones that don't match the username we want to remove. Finally, we write the filtered contacts back to the file.
```

While this is an advanced tutorial and thus probably actually straightforward, sometimes stating that things are straightforward can confuse the reader more. Because what if the reader is intermediate and didn't find it straightforward?
Either way, in this case it's redundant, because if the explanation is straightforward, might as well not mention it :P.

Author: > Either way, in this case it's redundant, because if the explanation is straightforward, might as well not mention it :P.

Lol, didn't see it like that, guess you are right :s

> Now is the turn of the search functions. The first one is the `searchId` function:

Reviewer: ```suggestion
Now onto the search functionality. The first one is the `searchId` function:
```

```scala 3
override def searchId(username: Username): IO[Option[Contact]] =
  for 
    contacts <- bookPath.readLines
  yield contacts
    .find(_.startsWith(username))

>     .flatMap(parseContact(_).toOption)

Reviewer: Why is it called `searchId` if it searches by `username`?

Author: In my mind, a username is unique in a database (just like an id). So, to avoid writing `searchByUsername` (quite lonmng) I  leaved it `searchId`. Should I change everywhere? :o

Reviewer 2: It wouldn't hurt. Let me be paternalistic for a moment: using a username, an email or any user submitted data as an id in a database is a terrible idea. Ids in general are meant to be unique and never changing (unlike a username), and as such they should be generated by the application's backend.

Reviewer 2: It's better to rename it as `searchUsername` and modify all the commands accordingly

The next three functions are very similar, so we will only show one of them, the `searchName` function:

```scala 3
override def searchName(name: Name): IO[List[Contact]] =
  for
    contacts <- bookPath.readLines.map(_.filter(_.contains(name)))
    parsed <- contacts.traverse(contact => 
      IO.fromEither(parseContact(contact)) 
    )
  yield parsed
```

The first part is similar to the previous functions, we load all the contacts from the file, and then we filter the ones that contain the name (or field) we want to search. Look! It's our friend `traverse` again! We're going to use it to parse the file representation of the contacts to our `Contact`using the `fromEither` method, as this will raise the error if it encounters a `Left` inside the `IO`.

Next, we have the `getAll` function to list all the contacts:

```scala 3
override def getAll: IO[List[Contact]] =
  for
    contacts <- bookPath.readLines
    parsed <- contacts.traverse(contact =>
      IO.fromEither(parseContact(contact))
    )
  yield parsed
```

> In like the search functions, we load all the contacts from the file, and then we parse them to a `Contact` using the `traverse` method.

Reviewer: ```suggestion
Just like the search functions, we load all the contacts from the file, and then we parse them to a `Contact` using the `traverse` method.
```
The wording here is also a bit iffy, `traverse` doesn't parse it, `parseContact` does it, but `traverse` makes it so that we return a `IO[List[Contact]]` and not a `List[IO[Contact]]`. I think it's important to convey that `traverse` actually does that.

Author: You are right!

```scala 3
override def updateContact(
    username: Username
)(modify: Contact => Contact): IO[Contact] =
  bookPath
    .readLines
    .flatMap(_.toVector.traverse(c => IO.fromEither(parseContact(c)))) // (1)
    .flatMap: contacts =>
      contacts.zipWithIndex.find(_._1.username == username) match      // (2)
        case Some((contact, index)) =>
          val updated = modify(contact)
          bookPath.writeLines(                                         // (3)              
            contacts.updated(index, updated).map(encodeContact)        // (4)
          ) *> IO.pure(updated)
        case None =>
          new Exception(s"Contact $username nor found")
            .raiseError[IO, Contact]
      
    
```

1. We convert all the contacts from the file by parsing them to a `Contact` using the `traverse` method along with the `fromEither` method. Note that we also transform the `List` to a `Vector`, this is because a Vector has better performance when we need to read or update elements in random positions. 

> 

Reviewer: Here also it reads as if `traverse` and `fromEither` do the parsing, but they actually don't, they make it more ergonomic to work with types.

2. After that, we will create a tuple containing the contact and its index in the list. We'll use this index to update the contact later.

3. If we find the contact, we will write the updated contact to the file using the `writeLines` method.

4. We update the contact in the list using the `updated` method that takes the index on which the element is located and the new element of the collection. This new element is created using the `modify` function of above. Finally, we map each of the contacts to a string using the `encodeContact` method, and then we write them to the file as we said in (3).

Oof, that was a lot of code! But we're done with the core of our application. Now we can move to the CLI part.

## CLI

> In this section, we will define the CLI part of our application. Here, we’re going to use the core login previously implemented to interact with the CLI. For example, by printing to the standard output and reading from the user input.  

Reviewer: ```suggestion
In this section, we will define the CLI part of our application. Here, we’re going to use the core logic previously implemented to interact with the CLI. For example, by printing to the standard output and reading from the user input.  
```

> It makes little sense to define the behaviour of the CLI in terms of an interface, as there are few ways of interacting with the user via a console. So we're going to contain all the logic in a class that has the `ContactManager` as a dependency.

Reviewer: ```suggestion
It makes little sense to define the behaviour of the CLI in terms of an interface, as there are only a few ways of interacting with the user via a console. So we're going to contain all the logic in a class that has the `ContactManager` as a dependency.
```

```scala 3
class Cli private (cm: ContactManager): 
  def addCommand: IO[Unit] = ...
```

The first function is going to be the `addCommand`. The idea is to ask the user step by step for the information of the contact and then add it to the contact list:

```scala 3
def addCommand: IO[Unit] =
  for 
    username    <- IO.println("Enter the username: ") >> IO.readLine
    firstName   <- IO.println("Enter the first name: ") >> IO.readLine
    lastName    <- IO.println("Enter the last name: ") >> IO.readLine
    phoneNumber <- IO.println("Enter the phone number: ") >> IO.readLine
    email       <- IO.println("Enter the email: ") >> IO.readLine

    contact = Contact(username, firstName, lastName, phoneNumber, email)
    
    _ <- cm.addContact(contact)
      .flatMap(username => IO.println(s"Contact $username added"))
      .handleErrorWith:
        case ContactFound(username) =>
          IO.println(s"Contact $username already exists")
        case e =>
          IO.println(s"An error occurred: \n${e.printStackTrace()}")
    
  yield ()
```

> The first five lines after the `for` asks the fields to save the contact. Then we create the `Contact` object with the information provided. Next, we call the `addContact` method of the `ContactManager` and handle the errors. If the contact already exists, we print a message to the user. If another error occurs, we print the stack trace of the error just to see what went wrong.

Reviewer: ```suggestion
The first five `IO.readLine` calls in `for` asks the fields to save the contact. Then we create the `Contact` object with the information provided. Next, we call the `addContact` method of the `ContactManager` and handle the errors. If the contact already exists, we print a message to the user. If another error occurs, we print the stack trace of the error to see what went wrong.
```

Now the `removeCommand` function:

```scala 3 
def removeCommand(username: Username): IO[Unit] =
  cm.removeContact(username) >> IO.println(s"Contact $username removed")
```

> It simply calls the `removeContact` function and print a message to the user. The function doesn't corroborate if the contact exists and will delete any matching username anyway, but you can add this logic if you want.

Reviewer: ```suggestion
It simply calls the `removeContact` function and prints a message to the user. The function doesn't check if the contact exists and will delete any matching username anyway, but you can add this logic if you want.
```

The `searchIdCommand` function is also straight forward, matches the `Options` returned by the `searchId` method and prints the user information if it exists:

```scala 3
def searchIdCommand(username: Username): IO[Unit] =
  for
    contact <- cm.searchId(username)
    _ <- contact match 
      case Some(c) => IO.println(c.show)
      case None    => IO.println(s"Contact $username not found")
  yield ()
```

Like the previous `searchX` variants, they’re fairly similar, so we will only show one of them.
In this case, the `searchEmailCommand` one:

```scala 3
def searchEmailCommand(email: Email): IO[Unit] =
  for {
    contacts <- cm.searchEmail(email)
    _        <- contacts.traverse_(c => IO.println(c.show))
} yield ()
``` 

It also uses the `traverse_` method, but with a little variant that optimises it a bit. It takes as a parameter a function from `A => IO[Unit]` and it will execute the effect for each element of the collection, but it will discard the result of the computation.

> Last but not least, the `updateCommand` function. The idea here is to take as parameters the username to change the contact and a list of flags parsed from the command line that contain information about which field to update and the new value of the field, and apply each of them:
> 

Reviewer: ```suggestion
Last but not least, the `updateCommand` function. The idea here is to take as parameters the username to change the contact and a list of flags parsed from the command line that contain information about which field to update and the new value of the field, and apply each of them:

```
This is a pretty long and hard to parse line, can you break the second sentence down a bit more, what are the parameters and to what does it apply each of them?

First thing is creating a parse function that will pattern match the `args` parameter of our entry point functions and, depending on the command, will call the respective function of the `Cli` class:

```scala 3
def parsePrompt(args: List[String]): CliCommand =
  args match 
    case Nil                                   => Help
    case "add" :: Nil                          => AddContact
    case "remove" :: username :: Nil           => RemoveContact(username)
    case "search" :: "id" :: username :: Nil   => SearchId(username)
    case "search" :: "name" :: name :: Nil     => SearchName(name)
    case "search" :: "email" :: email :: Nil   => SearchEmail(email)
    case "search" :: "number" :: number :: Nil => SearchNumber(number)
    case "list" :: _                           => ViewAll
    case "update" :: username :: options =>
      UpdateContact(username, parseUpdateFlags(options))
    case "--help" :: _ => Help
    case "help" :: _   => Help
    case _             => Help
```
Here we're using pattern matching to match the command and the arguments, so we can create the corresponding `CliCommand`.

> Notice that we als created a `parseUpdateFlags` function, this does a similar thing than the function from bellow, but it creates a list of flags that need to be updated in the contact:

Reviewer: ```suggestion
Notice that we also created a `parseUpdateFlags` function, this does a similar thing, but it creates a list of flags that can change the updating behaviour:
```

    exists <- path.exists
    _      <- dir.createDirectories.unlessA(exists)
    _      <- path.createFile.unlessA(exists)
  yield path
```
The `userHome` function will find the home directory and then, create the directory and the file if they don't exist.

After that, our main (`run`) function will be like this:

```scala 3
def run(args: List[String]): IO[ExitCode] =
  createBookPath.flatMap:
    bookPath =>
      val cm  = ContactManager.impl(bookPath)
      val cli = Cli.make(cm)
  
      Prompt.parsePrompt(args) match {
        case Help                    => cli.helpCommand
        case AddContact              => cli.addCommand
        case RemoveContact(username) => cli.removeCommand(username)
        case SearchId(username)      => cli.searchIdCommand(username)
        case SearchName(name)        => cli.searchNameCommand(name)
        case SearchEmail(email)      => cli.searchEmailCommand(email)
        case SearchNumber(number)    => cli.searchNumberCommand(number)
        case ViewAll                 => cli.viewAllCommand
        case UpdateContact(username, flags) =>
          cli.updateCommand(username, flags)
          
  .as(ExitCode.Success)
```

> After creating our working path,
> we start creating our dependencies by calling the methods inside the companion objects of our application, in this case, the contact manager and the CLI. Then we parse the command line arguments using the `parsePrompt` function and call the respective function of the `Cli` class.

Reviewer: ```suggestion
After creating our working path, we start creating our dependencies by calling the methods inside the companion objects of our application, in this case, the contact manager and the CLI. Then we parse the command line arguments using the `parsePrompt` function and call the respective function of the `Cli` class.
```

      _      <- path.createFile.unlessA(exists)
    yield path

  def run(args: List[String]): IO[ExitCode] =
    createBookPath.flatMap:
      bookPath =>
        val cm  = ContactManager.impl(bookPath)
        val cli = Cli.make(cm)

        Prompt.parsePrompt(args) match {
          case Help                    => cli.helpCommand
          case AddContact              => cli.addCommand
          case RemoveContact(username) => cli.removeCommand(username)
          case SearchId(username)      => cli.searchIdCommand(username)
          case SearchName(name)        => cli.searchNameCommand(name)
          case SearchEmail(email)      => cli.searchEmailCommand(email)
          case SearchNumber(number)    => cli.searchNumberCommand(number)
          case ViewAll                 => cli.viewAllCommand
          case UpdateContact(username, flags) =>
            cli.updateCommand(username, flags)
            
    .as(ExitCode.Success)
```

And that's it!
We’ve created a simple CLI application that manages contacts using Shellfish and Cats Effect.
We hope you enjoyed this tutorial and learned a lot about how to use Shellfish in a real-world application.

### Exercise

> We used a concrete path to create and store our contacts.

Reviewer: ```suggestion
We used a hardcoded path to store our contacts.
```

But what if we want to change the path to another location?
Or maybe have multiple locations with different contacts? 

Try to modify the application so the user can initialise and change the location of a contact book on a specific path and then use that path to store the contacts.

To do that, you can create two new commands, `init` and `change` that will create a new contact book and change the current contact book, respectively.

> You can also create new domain representations of that book,
> like `ContactBook(path: Path, contacts: List[Contacts])` and modify the `ContactManager` to use that representation.

Reviewer: ```suggestion
You can also create new domain representations of that book, like `ContactBook(path: Path, contacts: List[Contacts])` and modify the `ContactManager` to use that representation.
```

Good luck!

> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
>  
> 
> 
> 
> 
> 

Reviewer: ```suggestion
```

Now the `removeCommand` function:

`~/src/main/shellfish/contacts/cli/Cli.scala`
```scala 3 
def removeCommand(username: Username): IO[Unit] =
  cm.removeContact(username) >> IO.println(s"Contact $username removed")
```
It simply calls the `removeContact` function and prints a message to the user. The function doesn't check if the contact exists and will delete any matching username anyway, but you can add this logic if you want.

The `searchIdCommand` function is also straight forward, matches the `Options` returned by the `searchId` method and prints the user information if it exists:

`~/src/main/shellfish/contacts/cli/Cli.scala`
```scala 3
def searchIdCommand(username: Username): IO[Unit] =
  for
    contact <- cm.searchId(username)
    _ <- contact match 
      case Some(c) => IO.println(c.show)
      case None    => IO.println(s"Contact $username not found")
  yield ()
```

Like the previous `searchX` variants, they’re fairly similar, so we will only show one of them.
In this case, the `searchEmailCommand` one:

> `~/src/main/shellfish/contacts/cli/Cli.scala`
> ```scala 3
> def searchEmailCommand(email: Email): IO[Unit] =

Reviewer: I would recommend this style for the path specification
```suggestion
```scala 3
// src/main/shellfish/contacts/cli/Cli.scala
def searchEmailCommand(email: Email): IO[Unit] =
```

    case "--help" :: _ => Help
    case "help" :: _   => Help
    case _             => Help
```
Here we're using pattern matching to match the command and the arguments, so we can create the corresponding `CliCommand`.

Notice that we also created a `parseUpdateFlags` function, this does a similar thing, but it creates a list of flags that can change the updating behaviour:  

`~/src/main/shellfish/contacts/cli/Prompt.scala`
```scala 3
private def parseUpdateFlags(options: List[String]): List[Flag] = 

  @tailrec
  def tailParse(remaining: List[String], acc: List[Flag]): List[Flag] =
    remaining match 
  
      case Nil => acc
  
      case "--first-name" :: firstName :: tail =>
        tailParse(tail, FirstNameFlag(firstName) :: acc)
  
      case "--last-name" :: lastName :: tail =>
        tailParse(tail, LastNameFlag(lastName) :: acc)
  
      case "--phone-number" :: phoneNumber :: tail =>
        tailParse(tail, PhoneNumberFlag(phoneNumber) :: acc)
  
      case "--email" :: email :: tail =>
        tailParse(tail, EmailFlag(email) :: acc)
  
      case flag :: _ => List(UnknownFlag(flag))
    
  
  tailParse(options, Nil)
  
end parseUpdateFlags
```

> Because the update flags can come on any order, we use a recursive function to match each flag and its value, and then we create a list of `Flag` that will be passed to the `updateCommand` function like sayed before.

Reviewer: ```suggestion
Because the update flags  are unordered, we use a recursive function to match each flag and its value, and then we create a list of `Flag` that will be passed to the `updateCommand` function like defined before.
```

## Project Structure
For this project, we will use the following project structure:

```
examples/
└── src
    └── main
        └── scala
            └── contacts
                ├── app
                ├── cli
                ├── core
                └── domain
```

- `app`: This package will contain the main entry point of the application.
- `cli`: This package will contain the CLI logic such as prompt parsing and command execution.
- `core`: This package will contain the business logic of the application, here is where we will use Shellfish!
- `domain`: This package will contain important business logic (mostly types) that `app`, `cli` and `core` will use. 

Let us first start by defining our domain.

## Domain
Our domain will be straightforward, we will have a `Contact` case class that will represent a contact. 

```scala
//src/main/shellfish/contacts/domain/contact.scala
type Username    = String
type Name        = String
type PhoneNumber = String
type Email       = String

case class Contact(
  username: Username,
  firstName: Name,
  lastName: Name,
  phoneNumber: PhoneNumber,
  email: Email
)
``` 

> The type aliases are great, not only for added readability, but they also provide a way of not accidentally passing the wrong argument to the function. You can also use refined types provided by nice libraries like [Iron Types](https://github.com/Iltotore/iron), but for the sake of simplicity, we will use plain strings.

Reviewer 2: ```suggestion
The type aliases are great, not only for added readability, but they also provide a way of not accidentally passing the wrong argument to the function. You can also use refined types provided by libraries like [Iron](https://github.com/Iltotore/iron), but for the sake of simplicity, we will use plain strings.
```

> We will also need a function to display the contact in a nice format. If you came from Java, this function is known to `toString()`. But in Scala, we use `show` for this purpose.

Reviewer 2: ```suggestion
We will also need a function to display the contact in a nice format.
```

Using `show` is not a standard Scala thing (sadly)

```scala 3
//src/main/shellfish/contacts/domain/contact.scala
case class Contact( ... ):
  def show: String = s"""
    |----- $username -----
    |
    |First Name:    $firstName
    |Last Name:     $lastName
    |Phone:         $phoneNumber
    |Email:         $email
  """.stripMargin
```

> We will also define a custom error type that gets when a contact already exists.  

Reviewer 2: ```suggestion
We will also define a custom error type that gets thrown when adding an already existing contact.  
```

  
    private def parseContact(contact: String): IO[Contact] =
      contact.split('|') match 
        case Array(id, firstName, lastName, phoneNumber, email) =>
          Contact(id, firstName, lastName, phoneNumber, email).pure[IO]
        case _ =>
          new Exception(s"Invalid contact format: $contact")
            .raiseError[IO, Contact]
```

Next, a function to encode the contact to a string:

```scala 3
//src/main/shellfish/contacts/core/ContactManager.scala
    private def encodeContact(contact: Contact): String =
      s"${contact.username}|${contact.firstName}|${contact.lastName}|${contact.phoneNumber}|${contact.email}"
```

Finally, a helper function to help us save the contacts in memory:
```scala 3
    private def saveContacts(contacts: List[Contact]): IO[Unit] =
      bookPath.writeLines(contacts.map(encodeContact))
```

Now we will start with the `getAll` function. This is because our contact manager will work by first loading and parsing all the contacts so that we can perform operations on them:

```scala 3
override def getAll: IO[List[Contact]] = 
  for 
    lines    <- bookPath.readLines
    contacts <- lines.traverse(parseContact)
  yield contacts
```

> We first load all the contacts into memory using our `readLines` function and them parse them into the type `Contact`. But look! It's our friend `traverse` again! We're going to use it to convert our `List[IO[Contact]]` into a `IO[List[Contact]]`. Why? Because the `parseContacts` function already returns an `IO`, so it suits perfectly to the function that `traverse` needs as argument. 

Reviewer 2: ```suggestion
We first load all the contacts in memory using our `readLines` function and then parse them into `Contact`s. We're going to use `traverse` to convert our `List[String]` into a `IO[List[Contact]]`. Why? Because the `parseContacts` function returns an `IO[Contact]`, so it suits perfectly for our use case. 
```

> Now, we can implement the `addContact` function:

Reviewer 2: ```suggestion
We can now implement the `addContact` function:
```

```scala 3
//src/main/shellfish/contacts/core/ContactManager.scala
override def addContact(contact: Contact): IO[Username] = 
  for 
    contacts <- getAll                                      // (1)
    _ <- IO(contacts.contains(contact)).ifM( // (2)
      ContactFound(contact.username).raiseError[IO, Unit],
      saveContacts(contact :: contacts)
    )
  yield contact.username 
```

> First, we read all of our contacts, line by line using the `ContactManager#getAll` method (1). Then, we check if the contact already exists in the file checking if the username is contained (2). If it does, we raise an error using the `raiseError` method. If it doesn't, we save it to memory using our `saveContacts` methods. This is done with the `ifM`; this method basically checks if the boolean inside the `IO` is true or false and it executes an effect depending on the condition. 

Reviewer 2: ```suggestion
First, we read all of our contacts, line by line using the `ContactManager.getAll` method (1). Then, we check if the contact already exists in the file checking if the username is contained (2). If it does, we raise an error using the `raiseError` method. If it doesn't, we save it to memory using our `saveContacts` methods. This is done with the `ifM`; this method checks for a boolean value in an `IO` and executes one of the two  declared effects depending on it.
```

  yield ()
```

Here we load all the contacts from the file, and then we filter the ones that don't match the username we want to remove. Finally, we write the filtered contacts back to the file.

Now onto the search functionality. The first one is the `searchId` function:

```scala 3
//src/main/shellfish/contacts/core/ContactManager.scala
override def searchUsername(username: Username): IO[Option[Contact]] =
  getAll.map(contacts => contacts.find(_.username === username))
```

Again we load all the contacts from the file, and then we find the contact that matches the username we want to search using the `find` functions, it returns an `Option` whenever it found something or not. 

The next three functions are very similar, so we will only show one of them, the `searchName` function:

```scala 3
//src/main/shellfish/contacts/core/ContactManager.scala
override def searchName(name: Name): IO[List[Contact]] =
  getAll.map(contacts =>
    contacts.filter(c => c.firstName === name || c.lastName === name)
  )
```

The first part is similar to the previous functions, we load all the contacts from the file, and then we filter the ones that contain the name (or field) we want to search. 

> The last function we need to implement is the `updateContact` function, be aware that this is the most complex function to implement, so we will break it down into smaller parts.

Reviewer 2: ```suggestion
The last function we need to implement is the `updateContact` function. Be aware that this is the most complex function to implement, so we will break it down into smaller parts.
```

```scala 3
//src/main/shellfish/contacts/core/ContactManager.scala
override def updateContact(
    username: Username
)(modify: Contact => Contact): IO[Contact] = 
  for 
    oldContact <- searchUsername(username) match // (1)
      case None          => ContactNotFound(username).raiseError[IO, Contact]
      case Some(contact) => contact.pure[IO]

    updatedContacts = modify(oldContact) :: contacts.filterNot(_ == oldContact) // (2)
    _ <- saveContacts(updatedContacts)
  yield updatedContact
      
    
```

> 1. First, we'll find if the Contact we want to modify exists. If it doesnt, we raise a `ContactNotFound` error.

Reviewer 2: ```suggestion
1. First, we'll find if the Contact we want to modify exists. If it doesn't, we raise a `ContactNotFound` error.
```

2. In case the contact exists, we are going to apply to it the transform function and add it to the loaded list. While doing that, we delete the old contact via the `filterNot` method.

3. Finally, we save the new contact list to the memory.

> Oof, that was a lot of code! But we're done with the core of our application. Now we can move to the CLI part.

Reviewer 2: ```suggestion
Now that we're done with the core of our application, we can move to the CLI part.
```

## CLI

> In this section, we will define the CLI part of our application. Here, we’re going to use the core logic previously implemented to interact with the CLI. For example, by printing to the standard output and reading from the user input.  

Reviewer 2: ```suggestion
In this section, we will define our application's command line interface: it will accept commands and print the output to stdout.
```

> It makes little sense to define the behaviour of the CLI in terms of an interface, as there are only a few ways of interacting with the user via a console. That why we will just create some functions contained on a `Cli` package:

Reviewer 2: ```suggestion
It makes little sense to define the behaviour of the CLI in terms of an interface, as there are only a few ways of interacting with the user via a console. That's why we will just wrap these functionalities in some functions in a `Cli` object:
```

```scala 3
//src/main/shellfish/contacts/cli/Cli.scala
object Cli:
  def addCommand ...
```

> One way of passing the `ContactManager` dependency is to pass it as an argument to the function. However, it can be tedious to do this every time we want to use the functionality of our `Cli`, that's why Scala has an option to let the compiler do it for us! That's via the `given`/`using` clauses; if we want the compiler to "inject" the dependency, we just need to use the `using` keyword in the argument to the function, and it will be passed automatically as long as there is a given instance of our dependency in scope. This feature is also know as [contextual abstractions](https://docs.scala-lang.org/scala3/book/ca-context-parameters.html), We recommend that you take a closer look at these and how to use them in more detail.

Reviewer 2: ```suggestion
One way of passing the `ContactManager` dependency to these functions is to pass it as an argument. However, it can be tedious to do this every time we want to use the functionality of our `Cli`, that's why Scala has an option declare implicit paramaters via the `given`/`using` clauses. We just need to use the `using` keyword in front of any function's argument and as long as there is a given instance in scope at call site for every `using` clause, that value will be passed automatically. This feature is called [context parameters](https://docs.scala-lang.org/scala3/book/ca-context-parameters.html).
```

> That said, the first function is going to be the `addCommand`. The idea is to ask the user step by step for the information of the contact and then add it to the contact list:

Reviewer 2: ```suggestion
That said, the first function is going to be the `addCommand`. The idea is to ask the user the contact's information one step at a time and then add it to the contact list:
```

```scala 3
//src/main/shellfish/contacts/cli/Cli.scala
def addCommand(using cm: ContactManager): IO[Unit] =
  for 
    username    <- IO.println("Enter the username: ") >> IO.readLine
    firstName   <- IO.println("Enter the first name: ") >> IO.readLine
    lastName    <- IO.println("Enter the last name: ") >> IO.readLine
    phoneNumber <- IO.println("Enter the phone number: ") >> IO.readLine
    email       <- IO.println("Enter the email: ") >> IO.readLine

    contact = Contact(username, firstName, lastName, phoneNumber, email)
    
    _ <- cm.addContact(contact)
      .flatMap(username => IO.println(s"Contact $username added"))
      .handleErrorWith:
        case ContactFound(username) =>
          IO.println(s"Contact $username already exists")
        case e =>
          IO.println(s"An error occurred: \n${e.printStackTrace()}")

>     

Reviewer 2: ```suggestion
```

  yield ()
```

> The first five `IO.readLine` calls in `for` asks the fields to save the contact. Then we create the `Contact` object with the information provided. Next, we call the `addContact` method of the `ContactManager` and handle the errors. If the contact already exists, we print a message to the user. If another error occurs, we print the stack trace of the error to see what went wrong.  

Reviewer 2: ```suggestion
The first five `IO.readLine` calls in the `for` comprehension ask for the contact's detail, we then attempt to create the `Contact` object with the information provided  calling `addContact` method of the `ContactManager` and handle the errors. If the contact already exists, we print a message to the user, if another error occurs, we print the stack trace of the error to see what went wrong.  
```

Now the `removeCommand` function:

```scala 3 
//src/main/shellfish/contacts/cli/Cli.scala
def removeCommand(username: Username)(using cm: ContactManager): IO[Unit] =
  cm.removeContact(username) >> IO.println(s"Contact $username removed")
```
It simply calls the `removeContact` function and prints a message to the user. The function doesn't check if the contact exists and will delete any matching username anyway, but you can add this logic if you want.

The `searchIdCommand` function is also straight forward, matches the `Options` returned by the `searchId` method and prints the user information if it exists:

```scala 3
//src/main/shellfish/contacts/cli/Cli.scala
def searchIdCommand(username: Username)(using cm: ContactManager): IO[Unit] =
  for
    contact <- cm.searchId(username)
    _ <- contact match 
      case Some(c) => IO.println(c.show)
      case None    => IO.println(s"Contact $username not found")
  yield ()
```

> Like the previous `searchX` variants, they’re fairly similar, so we will only show one of them.

Reviewer 2: ```suggestion
All the `searchX` variants are fairly similar so we'll show just one of them:
```

> In this case, the `searchEmailCommand` one:

Reviewer 2: ```suggestion
```

```scala 3
//src/main/shellfish/contacts/cli/Cli.scala
def searchEmailCommand(email: Email)(using cm: ContactManager): IO[Unit] =
  for {
    contacts <- cm.searchEmail(email)
    _        <- contacts.traverse_(c => IO.println(c.show))
} yield ()
``` 

> It also uses the `traverse_` method, but with a little variant that optimises it a bit. It takes as a parameter a function from `A => IO[Unit]` and it will execute the effect for each element of the collection, but it will discard the result of the computation.

Reviewer 2: ```suggestion
`searchEmailCommand` uses the `traverse_` method: a little variant of `traverse` that takes a `A => IO[Unit]` function and then discards the result.
```

Last but not least, the `updateCommand` function. 

> We take as a first parameter the username of the contact we want to update in the database, so we can search for it's existence. 

Reviewer 2: ```suggestion
We take as a first parameter the username of the contact we want to update in the database, so we can search for its existence. 
```

As the second parameter, we will pass a list of flags, essentially describing which fields of the contact will be changed:

```scala 3
//src/main/shellfish/contacts/cli/Cli.scala
def updateCommand(username: Username, options: List[Flag])(
  using cm: ContactManager
): IO[Unit] =
  cm.updateContact(username): prev =>
    options.foldLeft(prev): (acc, flag) => // (1)
      flag match 
        case FirstNameFlag(name)     => acc.copy(firstName = name) // (2)
        case LastNameFlag(name)      => acc.copy(lastName = name)
        case PhoneNumberFlag(number) => acc.copy(phoneNumber = number)
        case EmailFlag(email)        => acc.copy(email = email)
        case UnknownFlag(_)          => acc
    
  .flatMap(c => IO.println(s"Updated contact ${c.username}")) // (3)
  .handleErrorWith:
    case ContactNotFound(username) =>
      IO.println(s"Contact $username not found")
    case e =>
      IO.println(s"An error occurred: \n${e.printStackTrace()}")
```

This is also quite a bit of code, but let's break it down: 

> First, we call the `updateContact` method of the `ContactManager` and pass the function that will modify the contact. To do this, we're going to reduce the list of flags using foldLeft. This function takes an initial value (the contact we want to update) and a function describing how to update the initial value (`acc`) with each one of the flags in turn (`flag`) (1). Where, we just use the `copy` methods and pattern match the flags to update the contact (2). After that, we print a message to the user saying that the contact was updated successfully (3). If the contact is not found, we print a message to the user saying that. 

Reviewer 2: ```suggestion
First, we call the `updateContact` method of the `ContactManager` and pass the function that will modify the contact. To do this, we're going to reduce the list of flags using `foldLeft`. This function takes an initial value (the contact we want to update) and a function describing how to update the initial value (`acc`) with each one of the flags in turn (`flag`) (1). Where, we just use the `copy` methods and pattern match the flags to update the contact (2). After that, we print a message to the user saying that the contact was updated successfully (3). If the contact is not found, we print a message to the user saying that. 
```

And as a last nice thing, a `helpCommand` function that will print the available commands to the user:

```scala 3
//src/main/shellfish/contacts/cli/Cli.scala
def helpCommand: IO[Unit] =
  IO.println:
    s"""
      |Usage: contacts [command]
      |
      |Commands:
      |  add
      |  remove <username>
      |  search id <username>
      |  search name <name>
      |  search email <email>
      |  search number <number>
      |  list
      |  update <username> [flags]
      |  help
      |
      |Flags (for update command):
      |  --first-name <name>
      |  --last-name <name>
      |  --phone-number <number>
      |  --email <email>
      |
      |""".stripMargin
```

> That was rough, but congratulations on making it so far! Now we can move to the last part of our application.

Reviewer 2: ```suggestion
That was tough, congratulations for making it so far. We can now move to the last part of our application.
```

## Prompt

> The last part is the prompt parsing, the functionality that is going to parse the user command line arguments into instructions that Scala can understand and then we can pass to the `Cli`.

Reviewer 2: ```suggestion
The last piece that we need to create is the prompt parsing, that will parse user submitted information into commands that our `Cli` can handle.
```

> We strongly suggest that you use a command parsing library, for example, [Decline](https://ben.kirw.in/decline/). 
> 
> But again, for the sake of simplicity, well demonstrate how we achieved that without the need of an external dependency. 

Reviewer 2: ```suggestion
We strongly suggest you to use a command line parsing library, like [Decline](https://ben.kirw.in/decline/) but for the sake of simplicity, we'll demonstrate how to create it without the need of an external dependency. 
```

> First thing is creating a parse function that will pattern match the `args` parameter of our entry point functions and, depending on the command, will call the respective function of the `Cli` class:

Reviewer 2: ```suggestion
The first thing we'll do is creating a `parsePrompt` function that will pattern match over the user input and, according to the value will return the appropriate `CliCommand`:
```

```scala 3
//src/main/shellfish/contacts/cli/Prompt.scala
def parsePrompt(args: List[String]): CliCommand =
  args match 
    case "add" :: Nil                          => AddContact
    case "remove" :: username :: Nil           => RemoveContact(username)
    case "search" :: "id" :: username :: Nil   => SearchId(username)
    case "search" :: "name" :: name :: Nil     => SearchName(name)
    case "search" :: "email" :: email :: Nil   => SearchEmail(email)
    case "search" :: "number" :: number :: Nil => SearchNumber(number)
    case "list" :: _                           => ViewAll
    case "update" :: username :: options =>
      UpdateContact(username, parseUpdateFlags(options))
    case Nil                                   => Help

>     case "--help" :: _ => Help
>     case "help" :: _   => Help
>     case _             => Help

Reviewer 2: ```suggestion
    case _             => Help
```

end parsePrompt
```

> Here we're using pattern matching to match the command and the arguments, so we can create the corresponding `CliCommand`.

Reviewer 2: ```suggestion
```

> Notice that we also created a `parseUpdateFlags` function, this does a similar thing, but it creates a list of flags that can change the updating behaviour:  

Reviewer 2: ```suggestion
We also created a `parseUpdateFlags` function that creates a list of flags that can change the updating behaviour using a similar pattern:  
```

```scala 3
//src/main/shellfish/contacts/cli/Prompt.scala`
private def parseUpdateFlags(options: List[String]): List[Flag] = 

  @tailrec
  def tailParse(remaining: List[String], acc: List[Flag]): List[Flag] =
    remaining match 
  
      case Nil => acc
  
      case "--first-name" :: firstName :: tail =>
        tailParse(tail, FirstNameFlag(firstName) :: acc)
  
      case "--last-name" :: lastName :: tail =>
        tailParse(tail, LastNameFlag(lastName) :: acc)
  
      case "--phone-number" :: phoneNumber :: tail =>
        tailParse(tail, PhoneNumberFlag(phoneNumber) :: acc)
  
      case "--email" :: email :: tail =>
        tailParse(tail, EmailFlag(email) :: acc)
  
      case flag :: _ => List(UnknownFlag(flag))
    
  
  tailParse(options, Nil)
  
end parseUpdateFlags
```

> Because the update flags  are unordered, we use a recursive function to match each flag and its value, and then we create a list of `Flag` that will be passed to the `updateCommand` function like defined before.

Reviewer 2: ```suggestion
Because the update flags can be unordered, we case use a recursive function to match each flag and its value and then create a list of `Flag`.
```

## Main

> Now that we have all the parts of our application, we can create the entry point of our application.

Reviewer 2: ```suggestion
Now that we have all the required components we can define the entry point of our application.
```

> The first thing is creating an initial function that describes a computation that will create the file we will work with. In this case, our in memory database will be located in `~/.shellfish/contacts.data`:

Reviewer 2: ```suggestion
The first thing that we need to do is create an initial function that that will create our in-memory database (located at `~/.shellfish/contacts.data`) if needed:
```

```scala 3
//src/main/shellfish/contacts/app/App.scala

> private val createBookPath: IO[Path] =

Reviewer 2: ```suggestion
private val getOrCreateBookPath: IO[Path] =
```

  for
    home <- userHome
    dir  = home / ".shellfish"
    path = dir / "contacts.data"
    exists <- path.exists
    _      <- dir.createDirectories.unlessA(exists)
    _      <- path.createFile.unlessA(exists)
  yield path
```

> The `userHome` function will find the home directory and then, create the directory and the file if they don't exist.

Reviewer 2: ```suggestion
```

> Remember the `using` keyword? Now we can create a given instance of the contact manager like this:
> 
> ```scala 3
> given cm: ContactManager = ContactManager(path)
> ```
> 
> With that, we can use the instance whenever we have declared an argument with the `using` clause with the same type.
> 
> Or even better, we can create it like this:

Reviewer 2: ```suggestion
We can then create our `given` instance of `ContactManager`:
```

```scala 3
createBookPath
  .map(ContactManager(_))

>   .flatMap: cm =>
>     given ContactManager = cm

Reviewer 2: ```suggestion
  .flatMap: 
    case given ContactManager =>
```

```

With that, our main (`run`) function will be like this:

```scala 3
//src/main/shellfish/contacts/app/App.scala
  def run(args: List[String]): IO[ExitCode] = createBookPath
    .map(ContactManager(_))

>     .flatMap: cm =>
>       given ContactManager = cm

Reviewer 2: ```suggestion
    .flatMap:
      case given ContactManager =>
```

      Prompt.parsePrompt(args) match 
        case Help                    => Cli.helpCommand
        case AddContact              => Cli.addCommand
        case RemoveContact(username) => Cli.removeCommand(username)
        case SearchId(username)      => Cli.searchUsernameCommand(username)
        case SearchName(name)        => Cli.searchNameCommand(name)
        case SearchEmail(email)      => Cli.searchEmailCommand(email)
        case SearchNumber(number)    => Cli.searchNumberCommand(number)
        case ViewAll                 => Cli.viewAllCommand
        case UpdateContact(username, flags) =>
          Cli.updateCommand(username, flags)
    
    .as(ExitCode.Success)
```

> 
> After creating our working path, we start creating our dependencies by calling the methods inside the companion objects of our application, in this case, the contact manager and the CLI. Then we parse the command line arguments using the `parsePrompt` function and call the respective function of the `Cli` class.  

Reviewer 2: ```suggestion

```

object App extends IOApp:

  private val createBookPath: IO[Path] =
    for
      home <- userHome
      dir  = home / ".shellfish"
      path = dir / "contacts.data"
      exists <- path.exists
      _      <- dir.createDirectories.unlessA(exists)
      _      <- path.createFile.unlessA(exists)
    yield path

  def run(args: List[String]): IO[ExitCode] =
    createBookPath.flatMap:
      bookPath =>
        val cm  = ContactManager.impl(bookPath)
        val cli = Cli.make(cm)

        Prompt.parsePrompt(args) match
          case Help                    => cli.helpCommand
          case AddContact              => cli.addCommand
          case RemoveContact(username) => cli.removeCommand(username)
          case SearchId(username)      => cli.searchIdCommand(username)
          case SearchName(name)        => cli.searchNameCommand(name)
          case SearchEmail(email)      => cli.searchEmailCommand(email)
          case SearchNumber(number)    => cli.searchNumberCommand(number)
          case ViewAll                 => cli.viewAllCommand
          case UpdateContact(username, flags) =>
            cli.updateCommand(username, flags)
            

>     .as(ExitCode.Success)

Reviewer 2: This is not up to date with the rest of the tutorial
